cub/test/catch2_test_warp_scan_partial.cu

```diff
@@ -0,0 +1,1287 @@
+  // undefined.
+
+  // When comparing device output, the corresponding undefined data points need
+  // to be fixed
+
+  for (size_t i = 0; i < h_exclusive_out.size(); i += logical_warp_threads)
+  {
+    d_exclusive_out[i] = h_exclusive_out[i];
+  }
+
+  REQUIRE(h_inclusive_out == d_inclusive_out);
+  REQUIRE(h_exclusive_out == d_exclusive_out);
+}
+
+C2H_TEST("Partial warp combination custom scan works with initial value", "[scan][warp]", logical_warp_threads)
+{
+  constexpr int logical_warp_threads = c2h::get<0, TestType>();
+  constexpr int total_warps          = total_warps_t<logical_warp_threads>::value();
+  constexpr int tile_size            = logical_warp_threads * total_warps;
+  using type                         = int;
+
+  const int valid_items = GENERATE_COPY(
+    take(1, random(2, cuda::std::max(2, logical_warp_threads))),
+    take(1, random(logical_warp_threads + 2, cuda::std::numeric_limits<int>::max())),
+    values({1, logical_warp_threads, logical_warp_threads + 1}));
+  const int bounded_valid_items = cuda::std::min(valid_items, logical_warp_threads);
+  const type filler =
+    GENERATE_COPY(take(1, random(cuda::std::numeric_limits<type>::lowest(), cuda::std::numeric_limits<type>::max())));
+  CAPTURE(valid_items, logical_warp_threads, filler, c2h::type_name<type>());
+  c2h::device_vector<type> d_inclusive_out(tile_size, thrust::no_init);
+  c2h::device_vector<type> d_exclusive_out(tile_size, thrust::no_init);
+  c2h::device_vector<type> d_in(tile_size, thrust::no_init);
+  c2h::gen(C2H_SEED(num_seeds), d_in);
+
+  const type initial_value = GENERATE_COPY(take(2, random(0, total_warps * logical_warp_threads)));
+  CAPTURE(d_in, initial_value);
+
+  warp_combine_scan<logical_warp_threads, total_warps>(
+    d_in, d_inclusive_out, d_exclusive_out, min_init_value_scan_op_t<type>{initial_value}, valid_items, filler);
+
+  c2h::host_vector<type> h_exclusive_out = d_in;
+  c2h::host_vector<type> h_inclusive_out = d_in;
+  for (size_t i = 0; i < tile_size; i += logical_warp_threads)
+  {
+    thrust::fill(
+      h_exclusive_out.begin() + i + bounded_valid_items, h_exclusive_out.begin() + i + logical_warp_threads, filler);
+    thrust::fill(
+      h_inclusive_out.begin() + i + bounded_valid_items, h_inclusive_out.begin() + i + logical_warp_threads, filler);
+  }
+
+  compute_host_reference(
+    scan_mode::exclusive,
+    h_exclusive_out,
+    logical_warp_threads,
+    [](type l, type r) {
+      return std::min(l, r);
+    },
+    valid_items,
+    initial_value);
+
+  compute_host_reference(
+    scan_mode::inclusive,
+    h_inclusive_out,
+    logical_warp_threads,
+    [](type l, type r) {
+      return std::min(l, r);
+    },
+    valid_items,
+    initial_value);
+
+  REQUIRE(h_inclusive_out == d_inclusive_out);
+  REQUIRE(h_exclusive_out == d_exclusive_out);
+}
+
+C2H_TEST(
+  "Partial warp scan does not apply op to invalid elements", "[scan][warp]", invalid_types, logical_warp_threads, modes)
```

Comment:
This test file got quite big again. I propose splitting off the "invalid elements" tests into another file and therefore putting the generalized kernels at the top of this file into a header.